### PR TITLE
MicroBit LightSensor GetAveragedValue

### PR DESCRIPTION
**Pull request by epeach**
Completes the MicroBit LightSensor API with getAveragedValue. Adds a buffer of 3 seconds of historical data and then allows students to request a period of time over which to average the data.

If the user calls getAveragedValue with a function under 50 or over 3000, an error is printed. Screenshot below:
Screenshot from 2020-08-20 10-54-39

Includes tests

## Links

- [spec](https://docs.google.com/document/d/1oPqJpIsaaGIQv7_KvCShmw6i0mT6MP3VqBP41yZ5M5c/edit#heading=h.ak61axyp722j)
- [jira](https://codedotorg.atlassian.net/browse/STAR-1152)

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
 import {EventEmitter} from 'events';
-import {sensor_channels, roundToHundredth} from './MicroBitConstants';
+import {
+  sensor_channels,
+  roundToHundredth,
+  SAMPLE_RATE,
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Maybe it's okay, but we seem to have three different naming conventions here.

File: apps/src/lib/kits/maker/boards/microBit/MicroBitConstants.js [hunk 2]
```diff
@@ -34,6 +34,12 @@ export const MB_SENSOR_VARS = [
   'compass'
 ];
 
+// milliseconds between samples for sensors
+export const SAMPLE_RATE = 50;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/MicroBitConstants.js]: "Sample rate" here would be how many times we sample per second (i.e. 1000 / 50 = 20).  What we appear to have here is actually SAMPLE_INTERVAL.

File: apps/src/lib/kits/maker/boards/microBit/MicroBitConstants.js [hunk 3]
```diff
@@ -34,6 +34,12 @@ export const MB_SENSOR_VARS = [
...
+
+// Max number of milliseconds for which we store historical data
+export const MAX_SENSOR_BUFFER = 3000;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/MicroBitConstants.js]: It feels like this variable name is missing a word indicating what this number means.  How about MAX_SENSOR_BUFFER_SIZE or MAX_SENSOR_SAMPLE_COUNT or something like that?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 4]
```diff
@@ -10,8 +15,19 @@ export default class LightSensor extends EventEmitter {
       rangeMax: 255,
       currentReading: 0
     };
+    // Track 3 seconds of historical data over which we can average sensor readings
+    this.buffer = new Float32Array(MAX_SENSOR_BUFFER / SAMPLE_RATE);
+    this.bufferIndex = 0;
+
     this.board = board;
     this.board.mb.addFirmataUpdateListener(() => {
+      //Record current reading to keep finite historical buffer
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: super nit: space after `//` and `.` at end of sentence.

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 5]
```diff
@@ -10,8 +15,19 @@ export default class LightSensor extends EventEmitter {
       rangeMax: 255,
       currentReading: 0
     };
+    // Track 3 seconds of historical data over which we can average sensor readings
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Might be worth commenting that this is a circular buffer.

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 6]
```diff
@@ -10,8 +15,19 @@ export default class LightSensor extends EventEmitter {
...
+    this.buffer = new Float32Array(MAX_SENSOR_BUFFER / SAMPLE_RATE);
+    this.bufferIndex = 0;
+
     this.board = board;
     this.board.mb.addFirmataUpdateListener(() => {
+      //Record current reading to keep finite historical buffer
+      this.buffer[this.bufferIndex] = this.board.mb.analogChannel[
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: I [noticed](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Float32Array) that this data type is stored in "platform byte order".  Does the assignment from this board interface into our buffer value always work, given this?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 7]
```diff
@@ -10,8 +15,19 @@ export default class LightSensor extends EventEmitter {
       rangeMax: 255,
       currentReading: 0
     };
+    // Track 3 seconds of historical data over which we can average sensor readings
+    this.buffer = new Float32Array(MAX_SENSOR_BUFFER / SAMPLE_RATE);
+    this.bufferIndex = 0;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Maybe `currentBufferWriteIndex` just to be really explicit about what this value is and what it's tracking?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 8]
```diff
@@ -67,11 +82,39 @@ export default class LightSensor extends EventEmitter {
...
+    }
+  }
 
   setRange(min, max) {
     this.state.rangeMin = min;
     this.state.rangeMax = max;
   }
 }
+
+function calculateValueWithinRange(value, min, max) {
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Can we have a comment explaining what this function does?  The function name is also a bit vague.. what exactly are we calculating?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 9]
```diff
@@ -67,11 +82,39 @@ export default class LightSensor extends EventEmitter {
...
+  let ratio = value / 255;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Why 255?  Can we put it into a named constant and explain why it's this value?

File: apps/src/lib/kits/maker/boards/microBit/MBFirmataWrapper.js [hunk 10]
```diff
@@ -12,7 +13,7 @@ export default class MicrobitFirmataWrapper extends MBFirmataClient {
     return Promise.resolve()
       .then(() => this.connect())
       .then(() => {
-        return this.setAnalogSamplingInterval(50);
+        return this.setAnalogSamplingInterval(SAMPLE_RATE);
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/MBFirmataWrapper.js]: Aha, as commented already, sampling rate is not the same as sampling interval :)

File: apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js [hunk 11]
```diff
@@ -70,6 +71,10 @@ export function cleanupMicroBitComponents(
       rangeMax: 255,
       currentReading: 0
     };
+    components.lightSensor.buffer = new Float32Array(
+      MAX_SENSOR_BUFFER / SAMPLE_RATE
+    );
+    components.lightSensor.bufferIndex = 0;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js]: How are these related to the similar buffer (and index) that are created in microBit/LightSensor.js?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 12]
```diff
@@ -67,11 +82,39 @@ export default class LightSensor extends EventEmitter {
     this.board.mb.stopStreamingAnalogChannel(sensor_channels.lightSensor); // disable light sensor
   }
 
-  //TODO
-  getAveragedValue() {}
+  // Get the averaged value over the give range of ms, adjusted within specified range
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: I see a given range (as the function parameter) but where is the specified range?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 13]
```diff
@@ -67,11 +82,39 @@ export default class LightSensor extends EventEmitter {
...
+  // TODO: Handle if the range is outside of 50 (before first sample) or 3000 (longer than our buffer).
+  // Currently returns null if range is outside of acceptable values
+  getAveragedValue(range) {
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: A "range" seems to imply a start and end point, but I think here we're really specifying a duration, which happens to end at the current moment..

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 14]
```diff
@@ -67,11 +82,39 @@ export default class LightSensor extends EventEmitter {
...
+      let indicesRange = Math.ceil(range / SAMPLE_RATE);
+      let sum = 0;
+      for (let i = 0; i < indicesRange; i++) {
+        // bufferIndex points to the next spot to write, so first historical
+        // data value is at (bufferIndex - 1)
+        let index =
+          (this.buffer.length + (this.bufferIndex - (i + 1))) %
+          this.buffer.length;
+        sum += this.buffer[index];
+      }
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: I found that the main assignment to `index` packed a few ideas into one line, and unexpectedly iterated backwards instead of forwards through the array (though for an average that's obviously okay).  Here's an attempt at unpacking things just a little bit:

```
const startIndex = this.bufferIndex - indicesRange;
const endIndex = this.bufferIndex - 1;

for (let index = startIndex; index <= endIndex; index++) {
  const actualIndex = (index + this.buffer.length) % this.buffer.length;
  sum += this.buffer[actualIndex];
}
```

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js (hunk 6 above)
Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Hi there! I don't know what this means. If you know, I'd love to chat about it, otherwise I'll do some research and see what I can find.

Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Discussed offline. Decided this implementation works.

File: apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js (hunk 11 above)
Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js]: They are the same - is called when the program is reset (via run/reset). So this is an opportunity for us to refresh the state of the component - new range, new 'current data', etc. 

Originally, I figured I could skip this step - since this is a circular buffer, we'll just keep looping over old data and don't need to 'zero it out' between runs. However, then I thought about a case where a student requests 3 seconds of data after only running their program for one second. In that case, if we've not 'zero'ed' out between runs, they might get data from their old previous runs of the program - which seems like a data bleed we don't want. But if we 'zero out' between runs, they'd get the value zero for the array values that are yet to be written, which feels like it would be more logical for students.

Comment (breville) [reply apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js]: Ah, that makes sense.  Agreed that resetting this data between runs is a good idea.

I'd suggest factoring the common code into a helper function in LightSensor.js, and which this code can then call.  That way, we don't have duplicated code, and this code doesn't have to "reach into" the internal data of lightSensor.

I'd also suggest only doing a `new` for the initial allocation of the buffer, and then resetting its contents to 0, perhaps using the new `.fill(0)` function.  I'm not sure of performance characteristics, but this feels better behaviour than making an additional memory allocation and relying on the garbage collector to clean things up.

Separately, if we are including 0s in the average, they will drag the value down unnecessarily, won't they?  It seems it would be better to average only the values we have written.  In that case, we would need to know how many values we have actually stored during the current run.  And in that case, we probably don't need to 0 anything out, we just need to track which array entries have valid values for the current run.

**Comment by epeach** [comment]
@breville PTAL, if you get a chance - thanks!

File: apps/src/lib/kits/maker/boards/microBit/MicroBitConstants.js [hunk 15]
```diff
@@ -34,6 +34,14 @@ export const MB_SENSOR_VARS = [
   'compass'
 ];
 
+// milliseconds between samples for sensors
+export const SAMPLE_INTERVAL = 50;
+
+// Max number of milliseconds for which we store historical data
+export const MAX_SENSOR_BUFFER_LENGTH = 3000;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/MicroBitConstants.js]: Ah, just realised it's duration, rather than buffer length, so could perhaps we could name accordingly?

File: apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js (hunk 11 above)
Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js]: To your last point - yes - I like it! I'm planning to implement it such that, if the user requests the average over 6 indices, but we've only recorded 3 indices, we'll return the average just over the 3 indices, rather than throwing an error or returning some other value.

Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js]: Alright, pushed a change to cover this - thanks!

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 16]
```diff
@@ -8,14 +15,25 @@ export default class LightSensor extends EventEmitter {
...
     };
+
     this.board = board;
     this.board.mb.addFirmataUpdateListener(() => {
+      // Record current reading to keep finite historical buffer.
+      // Modulo the index to loop to the beginning of the buffer when it exceeds array size.
+      this.state.buffer[
+        this.state.currentBufferWriteIndex % this.state.buffer.length
+      ] = this.board.mb.analogChannel[SENSOR_CHANNELS.lightSensor];
+      this.state.currentBufferWriteIndex++;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js:32]: Slightly concerned that this is unbounded.  How long until this value overflows?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 17]
```diff
@@ -60,18 +77,76 @@ export default class LightSensor extends EventEmitter {
...
-    this.board.mb.stopStreamingAnalogChannel(sensor_channels.lightSensor); // disable light sensor
+    this.board.mb.stopStreamingAnalogChannel(SENSOR_CHANNELS.lightSensor); // disable light sensor
+  }
+
+  // Reset the state to initial values between runs
+  reset() {
+    this.state = {
+      threshold: 128,
+      rangeMin: 0,
+      rangeMax: 255,
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Is this `MAX_LIGHT_SENSOR_VALUE`?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js (hunk 16 above)
Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js:32]: Gah! I was also thinking about this. I moved in this direction so I could simplify the check for whether the user is requesting more data than we've written. But it feels easier to add a boolean or counter to check for that case than to worry about this value overflowing, which feels reasonably possible ("I'm a student in class, I run my program but then I walk away from my computer to work with a partner without resetting my program"). I'll make that pivot.

Comment (breville) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js:32]: That said, it might take so long that we'll never hit it in reality... just needs a quick investigation/calculation.

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 18]
```diff
@@ -60,18 +77,76 @@ export default class LightSensor extends EventEmitter {
...
+    // Divide ms range by sample rate of sensor
+    let requestedRange = Math.ceil(ms / SAMPLE_INTERVAL);
+    let indicesRange = requestedRange;
+    // User requested average over greater range than is recorded, so average over all
+    // recorded data.
+    if (requestedRange >= this.state.currentBufferWriteIndex) {
+      indicesRange = this.state.currentBufferWriteIndex;
+    }
+    let endIndex =
+      this.state.currentBufferWriteIndex % this.state.buffer.length;
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Given that we do a `% this.state.buffer.length` inside the loop, we don't need to also do it here.  Also note that `endIndex` here is not actually the final index processed, but one past that, because it's actually a reference to the index that is yet to be written.  It would be possible to subtract 1 here, and make the `for` loop inclusive, if we wanted to adjust this.

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js [hunk 19]
```diff
@@ -60,18 +77,76 @@ export default class LightSensor extends EventEmitter {
...
+    );
+
+    // Divide ms range by sample rate of sensor
+    let requestedRange = Math.ceil(ms / SAMPLE_INTERVAL);
+    let indicesRange = requestedRange;
+    // User requested average over greater range than is recorded, so average over all
+    // recorded data.
+    if (requestedRange >= this.state.currentBufferWriteIndex) {
+      indicesRange = this.state.currentBufferWriteIndex;
+    }
```
Comment (breville) [inline apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: It's nice to keep values immutable if possible, so perhaps we could assign the default value of `requestedRange` in an `else` statement?

File: apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js [hunk 20]
```diff
@@ -60,11 +64,48 @@ describe('LightSensor', function() {
...
+        MAX_SENSOR_BUFFER_DURATION / SAMPLE_INTERVAL
+      );
+      for (let i = 0; i < lightSensor.state.buffer.length; i++) {
+        lightSensor.state.buffer[i] = 5;
+      }
+
+      lightSensor.state.currentBufferWriteIndex = 20;
+      expect(lightSensor.getAveragedValue(800)).to.equal(11.96);
     });
   });
```
Comment (breville) [inline apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js:129]: Do we have any tests for the case that we request the average of more samples than we currently have recorded?

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js (hunk 16 above)
Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js:32]: Google says MaxValue is 9007199254740992 in Javascript. If we're sampling every 50ms and incrementing this value then, then we'll hit max in 4.5035996e+14 seconds, which is 7.5059994e+12 minutes, which is many hours. So we're actually probably fine, since the value is incrementing only when the program is running.

Comment (breville) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js:32]: That sounds good!

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js (hunk 18 above)
Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Should for loops be inclusive? I noticed you also suggested that earlier in this code review. Is it a best practice?

File: apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js (hunk 20 above)
Comment (epeach) [reply apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js:129]: Yup! Covered in the 4th expect statement in "returns average of values in buffer".

File: apps/src/lib/kits/maker/boards/microBit/LightSensor.js (hunk 18 above)
Comment (breville) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: It comes down to the situation.  It's common when doing a 0-based count to use `<` in the condition , but in a situation like this, where we are very precisely iterating over an inclusive range, I think it's nice to use `<=` so that we can see exactly which indices we are processing.  If the condition were to compare against a `count` then `<` would read correctly, but given that we're talking about `endIndex` I think `<=` makes a bit more sense.

Comment (epeach) [reply apps/src/lib/kits/maker/boards/microBit/LightSensor.js]: Huh, cool. TIL. Updating . . .

File: apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js (hunk 20 above)
Comment (breville) [reply apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js:129]: Ah nice.  These cases are interesting enough that a little comment above each might be of interest.  Possibly we could also explain why 3000 is special, or just use the existing constant for it instead.

I actually ran the code manually to try some different cases and displayed each index that was being summed, because this sliding window stuff is tricky.  Definitely a case where upfront unit tests can help to validate everything is working as expected!

File: apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js [hunk 21]
```diff
@@ -60,11 +69,62 @@ describe('LightSensor', function() {
...
+        29.5
+      );
+
+      // Test that one index can be averaged
+      lightSensor.state.currentBufferWriteIndex = 1;
+      expect(lightSensor.getAveragedValue(SAMPLE_INTERVAL)).to.equal(0);
+
+      // Test the circular behavior of the buffer
+      lightSensor.state.currentBufferWriteIndex =
+        MAX_SENSOR_BUFFER_DURATION + 10;
```
Comment (breville) [inline apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js]: Should this be `MAX_SENSOR_BUFFER_DURATION / SAMPLE_INTERVAL + 10`?  It seems `MAX_SENSOR_BUFFER_DURATION` is on a different unit of measurement to the buffer itself...

File: apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js (hunk 20 above)
Comment (breville) [reply apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js:129]: Thanks for updating the test comments.  This coverage looks to be good.

I would consider these tests to be "integration tests" where we have added some coverage to prevent future regressions.  

As mentioned, I manually ran this code manually and hand-inspected every index that they consumed to make sure there were no off-by-one or other odd edge cases, which is very common for this type of code.  As a philosophical note, I'd recommend that, where possible, we make sure that unit testing for this kind of algorithm is set up so that we can very deliberately understand what the code is doing and we can easily hand-verify the result value that comes out.  Where we have a value like `11.96` as a result here, it will hopefully serve as protection against a future regression, but we are assuming that `11.96` is the appropriate result in the first place, and presumably got it as output from a previous run of the code, which is why I consider it more of an integration test than a pure upfront unit test, if that makes sense.

File: apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js (hunk 21 above)
Comment (epeach) [reply apps/test/unit/lib/kits/maker/boards/microBit/LightSensorTest.js]: It shouldn't change the behavior since this behavior gets mod'ed, but I'll update for clarity.

**Review by breville — APPROVED** [review]
Thanks for the iterations on this.  Feels like a great implementation!
